types/react/index.d.ts

```diff
@@ -1122,15 +1126,14 @@ declare namespace React {
     // A specific function type would not trigger implicit any.
     // See https://github.com/DefinitelyTyped/DefinitelyTyped/issues/52873#issuecomment-845806435 for a comparison between `Function` and more specific types.
     // tslint:disable-next-line ban-types
-    function useCallback<T extends Function>(callback: T, deps: DependencyList): T;
+    function useCallback<T extends Function>(callback: T, deps: DependencyList | void | null): T;
     /**
      * `useMemo` will only recompute the memoized value when one of the `deps` has changed.
      *
      * @version 16.8.0
      * @see https://react.dev/reference/react/useMemo
      */
-    // allow undefined, but don't make it optional as that is very likely a mistake
-    function useMemo<T>(factory: () => T, deps: DependencyList | undefined): T;
+    function useMemo<T>(factory: () => T, deps: DependencyList | void | null): T;
```

Comment:
We did not want to allow omitted `deps`. That would also trigger lint rules. I don't even understand why we allowed the `undefined`. If people want to disable lint rule they can do that with `eslint-disable-next-line`. And if they want to go around type-checking they have `@ts-expect-error`.

I think we should actually remove the `| undefined` here. The comment doesn't really help since it just reiterates the "what" but not the "why". I'll run some integration tests and git-blame and if that doesn't turn up anything we just make `deps` required for `useMemo`.

Reply:
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/67454

Reply:
In case of useMemo I think it only boils down to some kind of rare case of "functional programming" when someone would like to enable/disable memoization depending on some factors, without breaking rule of hooks in component.